tests/test_schemas.py

```diff
@@ -0,0 +1,303 @@
+            "sources": [
+                {"purl": "pkg:pypi/six@1.15.0"},
+            ],
+        }
+    ],
+    "content_sets": ["nodejs-rpms", "extra-rpms"],
+    "metadata": {
+        "icm_spec": "https://raw.githubusercontent.com/containerbuildsystem/atomic-reactor/f4abcfdaf8247a6b074f94fa84f3846f82d781c6/atomic_reactor/schemas/content_manifest.json", # noqa
+        "icm_version": 1,
+        "image_layer_index": -1,
+    },
+}
+CONTENT_MANIFEST_MISSING_SOURCES_PURL = {
+    "image_contents": [
+        {
+            "dependencies": [
+                {"purl": "pkg:pypi/six@1.15.0"},
+            ],
+            "purl": "pkg:generic/osbs/cachito-pip-with-deps?vcs_url=https%3A%2F%2Fgitlab.cee.redhat.com%2Fosbs%2Fosbs-test-cachito-project.git%40b3b2684a42971a6a7afc2b88106d908920519512", # noqa
+            "sources": [],
+        }
+    ],
+    "content_sets": ["nodejs-rpms", "extra-rpms"],
+    "metadata": {
+        "icm_spec": "https://raw.githubusercontent.com/content_manifest.json",
+        "icm_version": 1,
+        "image_layer_index": -1,
+    },
+}
+
+
+@pytest.mark.parametrize(
+    "data, schema, err_message",
+    [
+        # source_containers_user_params.json
+        ({}, "schemas/source_containers_user_params.json", "validating 'required' has failed"),
+        (SOURCE_CONTAINERS_USER_PARAMS_MINIMAL, "schemas/source_containers_user_params.json", None),
+        (SOURCE_CONTAINERS_USER_PARAMS_ALL_PROPERTIES,
+         "schemas/source_containers_user_params.json", None),
+        (SOURCE_CONTAINERS_USER_PARAMS_MISSING_USER,
+         "schemas/source_containers_user_params.json",
+         r"validating 'required' has failed \('user' is a required property\)"),
+        (SOURCE_CONTAINERS_USER_PARAMS_MISSING_BUILD_JSON_DIR,
+         "schemas/source_containers_user_params.json",
+         r"validating 'required' has failed \('build_json_dir' is a required property\)"),
+        (SOURCE_CONTAINERS_USER_PARAMS_ADDITIONAL_NONDEFINED_PROPERTY,
+         "schemas/source_containers_user_params.json", None),  # TODO should this fail?
+
+        # user_params.json
+        (USER_PARAMS_MINIMAL, "schemas/user_params.json", None),
+        (USER_PARAMS_MISSING_BUILD_JSON_DIR,
+         "schemas/user_params.json",
+         r"validating 'required' has failed \('build_json_dir' is a required property\)"),
+        (USER_PARAMS_MISSING_USER,
+         "schemas/user_params.json",
+         r"validating 'required' has failed \('user' is a required property\)"),
+        (USER_PARAMS_MISSING_GIT_URI,
+         "schemas/user_params.json",
+         r"validating 'required' has failed \('git_uri' is a required property\)"),
+        (USER_PARAMS_MISSING_GIT_REF,
+         "schemas/user_params.json",
+         r"validating 'required' has failed \('git_ref' is a required property\)"),
+        (USER_PARAMS_ADDITIONAL_NONDEFINED_PROPERTY,
+         "schemas/user_params.json", None),  # TODO should this fail?
```

Comment:
Probably should, but if we change it now, we potentially break existing builds :disappointed:

Reply:
This is maybe not an big issue and can probably be left as is, I'll keep this conversation unresolved in case anybody else wants to pitch in.